### DISCUSSION TITLE
Path instalation for CustomWidgets??

### DISCUSSION THREAD

**Discussion by esbvaGames** [alejandroautalan/pygubu-designer#196] [General]
I have an app in a directory, and inside this I have another folder "myWidgets", I can read and use them from pygubu, but they don't really work for me there, because when the app is compiled with pyinstaller, pygubu doesn't find them and the app, hangs, ui interface, if seen in designer, but custom widgets don't run when exported with pyinstaller or run main app module with python. Where do the custom-widgets actually need to be installed in order for them to be recognized and run, just like the designer defaults?
![WFUPSx5IgA](https://user-images.githubusercontent.com/5530499/222585222-39bc32f8-3bbd-43f1-98c8-edfdc31c2b78.png)
![eDxn8EZiXI](https://user-images.githubusercontent.com/5530499/222585264-1aee910c-e32e-4c00-bc7f-5c0bfad29ad6.png)

**Comment by esbvaGames** [comment]
![36NBDu8k8O](https://user-images.githubusercontent.com/5530499/222621202-56907a62-a8b6-4f1e-b6d6-30f1f072a89b.png)

**Reply by esbvaGames** [reply]
Command for build pyinstaller.
@echo.
pyinstaller AppBodega.py --collect-submodules tablebuttons.py --collect-submodules tablebuttonsapp.py --add-data AppBodega.ui;.\ --add-data .\Custom;.\Custom --path .\Custom --workpath .\binary\build --distpath .\binary  --noconfirm --onefile
IF ERRORLEVEL 1 pause

**Comment by alejandroautalan** [comment]
Hello @esbvaGames, thanks for trying pygubu.

I understand the problem. The solution is a little complicated but here we go.

I will assume this reduced project example, similar to yours.

```
../AppBodega/                    <- Your project directory
├── bodega_designer_helper.py    <- New helper for pygubu-designer
├── bodegamainapp.py             <- Your Main Application file.
├── bodega_main.ui               <- ui file
├── custom                       <- Your custom widgets dir, now a module
│   ├── __init__.py              <- Empty file to convert directory in module
│   ├── timerappwidgets.py       <- time widget builder
│   └── timewidget.py            <- time widget
└── response.txt
```

First you need to convert your custom widget directory in a module.
Create a empty file named \_\_init\_\_.py in custom directory.

Now your custom widgets live in a module, so the files will need modification:
Change the timerappwidgets.py file to this (notice the . before timewidget):

```python
# timerappwidgets.py
from pygubu.api.v1 import BuilderObject, register_widget
from .timewidget import TimeWidget
```

Next. Now that the builders live in a module you can not add them 
in Preferences > Custom Widgets.
We will use the file bodega_designer_helper.py to load the custom widgets
in pygubu-designer.
The content of this file is the following:

```python
# bodega_designer_helper.py

import custom.timerappwidgets
# import custom.your_other_custom_widgets_module
```

Now you can load this file in Designer > Preferences > Custom Widgets.
Clear any other files referencing your custom widgets.

![2023-03-03_00-29](https://user-images.githubusercontent.com/3482471/222625107-f23f8fe0-c252-4361-8780-fb9962a111ca.png)

Restart pygubu-designer and you should see your custom widgets again.

![2023-03-03_00-29_1](https://user-images.githubusercontent.com/3482471/222625143-e95a4ef0-d0df-4c80-97b5-f7e6fb8db4f7.png)

Next. In your application start file (in this case bodegamainapp.py)
you should also load the custom widgets.
Add this in your main app file:

```python
# bodegamainapp.py

import custom.timerappwidgets
# import custom.your_other_custom_widgets_module

...
```

Now, run the main app file and it should run correctly:

![2023-03-03_00-31](https://user-images.githubusercontent.com/3482471/222625430-90c573ee-cc63-48d5-bccf-1febef1120d9.png)

The test project here:

[AppBodega.zip](https://github.com/alejandroautalan/pygubu-designer/files/10877984/AppBodega.zip)


Try it, and let me know.
Regards!

Alejandro A.

**Reply by esbvaGames** [reply]
Ok, thank you very much, I'm going to try all this tomorrow, you saved my day 🤗🤗😅

**Comment by esbvaGames** [comment]
Thank you very much if I could verify it, all the problems were solved xon pyInstaller, also add:
import sys
sys.path.append("./custom")
at the beginning, and also the importers of the declarations of the pygubu-designer widgets, here I leave you a video, as a summary !!!

[https://youtu.be/HLGasi3NCv0](url)

**Reply by alejandroautalan** [reply]
Hello. I see that you found a more simple solution, great!
Also thanks for sharing the video showing the full process.

I'm really glad that pygubu is useful to you.
Regards!

Alejandro A.
